compiler/modules/front/codeGen.bal

```diff
@@ -1889,40 +1889,56 @@ function typedOperand(bir:Operand operand) returns TypedOperand? {
 }
 
 function operandPairOrderType(CodeGenContext cx, bir:Operand left, bir:Operand right) returns bir:OrderType? {
-    if operandIsNil(left) {
-        return promoteToOptOrderType(operandOrderType(cx, right));
-    }
-    if operandIsNil(right) {
-        return promoteToOptOrderType(operandOrderType(cx, left));
-    }
     bir:OrderType? lot = operandOrderType(cx, left);
     bir:OrderType? rot = operandOrderType(cx, right);
     if lot == rot {
         return lot;
     }
-    lot = promoteToOptOrderType(lot);
-    rot = promoteToOptOrderType(rot);
+    [lot, rot] = promoteToMatchingOptOrderType(lot, rot);
     if lot == rot {
         return lot;
     }
     return ();
 }
 
-function promoteToOptOrderType(bir:OrderType? ot) returns bir:OrderType? {
-    if ot is bir:UniformOrderType {
-        return { opt: ot };
+function promoteToMatchingOptOrderType(bir:OrderType? lot, bir:OrderType? rot) returns [bir:OrderType?, bir:OrderType?] {
+    if rot == bir:nilOrderType {
+        bir:OrderType? lOpt = promoteToOptOrderType(lot);
+        return [lOpt, lOpt];
     }
-    else {
-        return ot;
+    if lot == bir:nilOrderType {
+        bir:OrderType? rOpt = promoteToOptOrderType(rot);
+        return [rOpt, rOpt];
+    }
+    // Both are bir:nilOrderType case handled before
```

Comment:
Elsewhere in the code I have dealt with this sort of situation using the following pattern:

```
match [lot is bir:OptOrderType, rot is bir:OptOrderType] {
    [true, true] => {

    }
    [true, false] => {

    }
}
```

Reply:
That looks nice, but I am relaying on the type narrowing here.

Reply:
I worked out by mapping `false` to `nil`. BTW why can't we have `typed-binding-pattern` in `match-pattern`?

Reply:
Actually, plain pattern matching worked.